### PR TITLE
Make the OAuth APIs in .NET testable

### PR DESCRIPTION
**Pull request by Jeffders**
This PR includes a handful changes to make testing bots using the OAuth APIs and OAuthPrompt easier. 

These include:
1. Introduce the IUserTokenProvider interface which includes all of the OAuth APIs on the BotFrameworkAdapter
2. Have BotFrameworkAdapter implement this interface
3. Make all of these methods on BotFrameworkAdapter virtual (so they can be overridden by mock frameworks).
4. Change OAuthPrompt to look for adapter's that implement IUserTokenProvider instead of BotFrameworkAdapter

Additionally, you can now write unit tests for your bot using the TestAdapter and TestFlow:

1. TestAdapter includes a mock, in-memory implementation of IUserTokenProvider
2. Added unit tests for this in-memory implementation
3. Added unit tests for OAuthPrompt that test the entire login flow both with and without a magic code response.

### CODE REVIEW

**Comment by tomlm** [comment]
Hey jeff, can we make sure this is applied to js side too?

File: libraries/Microsoft.Bot.Builder/IUserTokenProvider.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/IUserTokenProvider.cs]: This interface can't possibly need all of these using statments. Could you remove & sort?

File: tests/Microsoft.Bot.Builder.Dialogs.Tests/OAuthPromptTests.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Dialogs.Tests/OAuthPromptTests.cs]: New tests classes give me a warm fuzzy feeling. :)

Comment (Jeffders) [reply tests/Microsoft.Bot.Builder.Dialogs.Tests/OAuthPromptTests.cs]: Fixed!

File: libraries/Microsoft.Bot.Builder/IUserTokenProvider.cs (hunk 1 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/IUserTokenProvider.cs]: fixed!

**Pull request merged**